avail-server/src/main/kotlin/com/avail/server/io/files/AvailServerTextFile.kt

```diff
@@ -0,0 +1,319 @@
+	 * @return The [TracedAction] that preserves this edit and how to reverse
+	 *   it.
+	 */
+	override fun removeRange(
+		start: Int, end: Int, timestamp: Long): TracedAction
+	{
+		// The text to remove from the file. The offset for one-based indexing
+		// requires we begin removing from 1 beyond start.
+		val removed =
+			content.copyStringFromToCanDestroy(start + 1, end, false)
+
+		// The client will use 0-based indexing in its request, so we must
+		// adjust by 1. Here using the start as inclusive is good enough to
+		// offset by 1.
+		val first =
+			content.copyTupleFromToCanDestroy(1, start, false)
+
+		// As stated above, must add 1 to end to keep from preserving the last
+		// character marked for removal.
+		val second = content.copyTupleFromToCanDestroy(
+			end + 1, content.tupleSize(), false)
+		content = Casts.cast(
+			first.concatenateWith(second, false))
+		return TracedAction(
+			timestamp,
+			RemoveRange(start, end),
+			Insert(
+				removed.asNativeString().toByteArray(Charsets.UTF_16BE),
+				start))
+	}
+
+	/**
+	 * Insert the [ByteArray] data into the file at the specified location. This
+	 * should remove existing data in the file in this range and replace it
+	 * with the provided data. This should preserve all data outside of this
+	 * range. This is equivalent to calling [removeRange] with `start` and `end`
+	 * then calling [insert] with `data` and `start` as inputs.
+	 *
+	 * The client will use 0-based indexing in its request, so we must adjust by
+	 * one. The file, in zero-based indexing must be prefixed before the
+	 * requested `start`. Shifting to Avail's one-based indexing and utilizing
+	 * the [A_Tuple.copyTupleFromToCanDestroy]'s inclusive range requires the
+	 * remove being after the requested `start`. Because the `start` is shifted
+	 * and correctly inclusive in the first half of the file, the 2nd half of
+	 * the file, the `end` must begin at the index one place beyond `end`
+	 * (`end + 1`). Thus the inserted text happens after the `start` position.
+	 *
+	 * @param data
+	 *   The `ByteArray` data to add to this [AvailServerFile].
+	 * @param start
+	 *   The location in the file to inserting/overwriting the data, exclusive.
+	 * @param end
+	 *   The location in the file to stop overwriting. All data after this point
+	 *   should be preserved.
+	 * @param timestamp
+	 *   The time in milliseconds since the Unix Epoch UTC the update occurred.
+	 * @return The [TracedAction] that preserves this edit and how to reverse
+	 *   it.
+	 */
+	override fun insertRange(
+		data: ByteArray, start: Int, end: Int, timestamp: Long): TracedAction
+	{
+		// The text to insert in the file
+		val text =
+			StringDescriptor.stringWithSurrogatesFrom(
+				String(data, Charsets.UTF_16BE))
+
+		// The text to remove from the file. The offset for one-based indexing
+		// requires we begin removing from 1 beyond start.
+		val removed =
+			content.copyStringFromToCanDestroy(start + 1, end, false)
```

Comment:
We should ask Mark; he fed me that line of code.

Reply:
It should remain false because there are two copyStringFromToCanDestroy calls on the same string.  If it were true for the first one, the operation could chop up a tree tuple however it saw fit, which could make the second operation fail.  Conversely, there's no point making the second one true, because it's likely that the first operation would have made the original tuple immutable anyhow.